tests/utilities/test_apply_func.py

```diff
@@ -115,24 +150,54 @@ class ModelExample:
+            actual, type
+        ), "Reduction of a dataclass should result in a dataclass"
+        assert dataclasses.is_dataclass(actual.feature) and not isinstance(
+            actual.feature, type
+        ), "Reduction of a nested dataclass should result in a nested dataclass"
+        assert (
+            actual.example_ids == expected.example_ids
+        ), "Reduction of a nested dataclass did not yield the desired result"
+        assert torch.allclose(
+            actual.label, expected.label
+        ), "Reduction of a nested dataclass did not yield the desired result"
+        assert torch.allclose(
+            actual.feature.input_ids, expected.feature.input_ids
+        ), "Reduction of a nested dataclass did not yield the desired result"
+        assert np.allclose(
+            actual.feature.segment_ids, expected.feature.segment_ids
+        ), "Reduction of a nested dataclass did not yield the desired result"
+
+    _assert_nested_dataclass_reduction(reduced["i"], expected_result["i"])
+
+    assert dataclasses.is_dataclass(reduced["j"]) and not isinstance(
+        reduced["j"], type
+    ), "Reduction of a dataclass with a class var should result in a dataclass"
+    assert WithClassVar.class_var == 0, "Reduction of a dataclass with a class var should not change the class var"
     assert torch.allclose(
-        reduced["i"].label, expected_result["i"].label
-    ), "Reduction of a nested dataclass did not yield the desired result"
+        reduced["j"].dummy, expected_result["j"].dummy
+    ), "Reduction of a dataclass with a class var did not yield the desired result"
+
+    assert dataclasses.is_dataclass(reduced["k"]) and not isinstance(
+        reduced["k"], type
+    ), "Reduction of a dataclass with an init-only var should result in a dataclass"
     assert torch.allclose(
-        reduced["i"].feature.input_ids, expected_result["i"].feature.input_ids
-    ), "Reduction of a nested dataclass did not yield the desired result"
-    assert np.allclose(
-        reduced["i"].feature.segment_ids, expected_result["i"].feature.segment_ids
-    ), "Reduction of a nested dataclass did not yield the desired result"
+        reduced["k"].dummy, expected_result["k"].dummy
+    ), "Reduction of a dataclass with an init-only var did not yield the desired result"
+
+    assert dataclasses.is_dataclass(reduced["l"]) and not isinstance(
+        reduced["l"], type
+    ), "Reduction of a dataclass with class and init-only vars should result in a dataclass"
+    assert torch.equal(
+        WithClassAndInitVar.class_var, torch.tensor(0)
+    ), "Reduction of a dataclass with class and init-only vars should not change the class var"
+    try:
+        _assert_nested_dataclass_reduction(reduced["l"].dummy, expected_result["l"].dummy)
+    except AssertionError:
+        raise AssertionError("Reduction of a dataclass with class and init-only vars did not yield the desired result")
```

Comment:
```suggestion
    _assert_nested_dataclass_reduction(reduced["l"].dummy, expected_result["l"].dummy)
```
what's the point of that? all the messages there get discarded for this?

Reply:
The main intention was to highlight that the test failed for a dataclass with class and init vars, presuming that if it had gotten this far it must have succeeded for those dataclasses without.
I introduced the `_assert_nested_dataclass_reduction` to avoid duplicating the preexisting tests.

I agree it was a bit awkward and have refactored the dataclass testing mechanisms to minimize duplication and encapsulate dataclass comparison in `__eq__` methods.